### PR TITLE
2nd step toward  Investigate and propose package repositories API with similar core interface to packages API. #3496 

### PR DESCRIPTION
**Pull request by gfichtenholt**
added GetPackageRepositoryDetail() API as outlined in the spec.

### CODE REVIEW

**Comment by absoludity** [comment]
Hi Greg. Thanks for the nice focused change again :) Note that the changes to the auto-generated files appear to be a bit excessive (in the dashboard) which probably means you need to run `yarn prettier` in the dashboard directory (until we automate the generation, each time you `buf generate`, you'll need to go to the `dashboard` directory and `yarn prettier`).

**Comment by gfichtenholt** [comment]
> Hi Greg. Thanks for the nice focused change again :) Note that the changes to the auto-generated files appear to be a bit excessive (in the dashboard) which probably means you need to run `yarn prettier` in the dashboard directory (until we automate the generation, each time you `buf generate`, you'll need to go to the `dashboard` directory and `yarn prettier`).

I did run 'yarn prettier' before pushing. Let me know which I will manually remove any dashboard-related changes from this PR. Thanks

**Comment by absoludity** [comment]
> I did run 'yarn prettier' before pushing. Let me know which I will manually remove any dashboard-related changes from this PR. Thanks

Hmm, could be actual changes then. I'll check locally later and let you know.

File: cmd/kubeapps-apis/core/packages/v1alpha1/repositories_test.go [hunk 1]
```diff
@@ -102,3 +132,78 @@ func TestAddPackageRepository(t *testing.T) {
...
+					Identifier: "repo-1",
+					Plugin:     mockedPackagingPlugin1.plugin,
+				},
+			},
+			expectedResponse: &corev1.GetPackageRepositoryDetailResponse{
+				Detail: plugin_test.MakePackageRepositoryDetail("repo-1", mockedRepoPlugin1.plugin),
+			},
+			statusCode: codes.OK,
+		},
+		/*
```
Comment (absoludity) [inline cmd/kubeapps-apis/core/packages/v1alpha1/repositories_test.go:165]: Is this commented as a reminder TODO later, or did you mean to delete it? The test mock that you modified below could return a NotFound easily enough?

File: cmd/kubeapps-apis/plugin_test/utils.go [hunk 2]
```diff
@@ -153,3 +160,23 @@ func MakePackageAppVersion(appVersion, pkgVersion string) *corev1.PackageAppVers
...
+		NamespaceScoped: false,
+		Type:            "helm",
+		Url:             DefaultRepoURL,
+		Interval:        DefaultRepoInterval,
+		TlsConfig:       nil,
+		Auth:            nil,
+		CustomDetail:    nil,
+		Status:          defaultRepoStatus,
+	}
+}
```
Comment (absoludity) [inline cmd/kubeapps-apis/plugin_test/utils.go:182]: Thanks for doing the tests here following the existing pattern (using `Make*` factory methods etc.)

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go [hunk 3]
```diff
@@ -81,7 +95,22 @@ func (s *Server) availableChartDetail(ctx context.Context, repoName types.Namesp
...
-	return availablePackageDetailFromChartDetail(chartID, chartDetail)
+	pkgDetail, err := availablePackageDetailFromChartDetail(chartID, chartDetail)
+	if err != nil {
+		return nil, err
+	}
+
+	// fix up a couple of fields that don't come from the chart tarball
+	repoUrl := repo.Spec.URL
+	if repoUrl == "" {
+		return nil, status.Errorf(codes.NotFound, "Missing required field spec.url on repository %q", repoName)
```
Comment (absoludity) [inline cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go:106]: Hmmm, not sure about a not found here? With helm you can install a chart from a repo, then later delete the repo. Is that same situation impossible with flux? Or do we return a not found in helm?

**Review by absoludity — APPROVED** [review]
Great, thanks Greg. I can prettier locally and don't see changes, so my mistake: just but generated changes.

File: cmd/kubeapps-apis/core/packages/v1alpha1/repositories_test.go (hunk 1 above)
Comment (gfichtenholt) [reply cmd/kubeapps-apis/core/packages/v1alpha1/repositories_test.go:165]: whooops. Started and forgot about it :-) Will fix

File: cmd/kubeapps-apis/plugin_test/utils.go (hunk 2 above)
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugin_test/utils.go:182]: my pleasure :-)

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go (hunk 3 above)
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go:106]: This was here a long time, I just moved the code around a bit. I think it was for some negative test I wrote. Anyway, `repo.Spec.URL` is a required field in flux https://fluxcd.io/docs/components/source/helmrepositories/ so in theory the if clause never triggers under normal circumstances

Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go:106]: regarding your question about fix. The helm chart CR has a back reference to the source. So I don't think this situation is possible. If you delete the helm repo, then associated helmcharts are deleted too. But I will verify now just to make sure

Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go:106]: I stand corrected. It is possible to delete the helmrepository. Then the helm chart goes into a 'bad' (not ready) state:
```
[v1alpha1]$ kubectl get helmchart
NAME                 CHART     VERSION   SOURCE KIND      SOURCE NAME   READY   STATUS                                         AGE
default-my-podinfo   podinfo   *         HelmRepository   podinfo       True    Pulled 'podinfo' chart with version '6.1.0'.   14s
[v1alpha1]$ kubectl get helmrepository
NAME      URL                                      READY   STATUS                                                                               AGE
podinfo   https://stefanprodan.github.io/podinfo   True    Fetched revision: 4995e6e5e300c68f9ea8e0f26c91e42f0dbaf1be46e4e464b25988eb5ae86d4b   97s
[v1alpha1]$ kubectl delete helmrepository/podinfo
helmrepository.source.toolkit.fluxcd.io "podinfo" deleted
[v1alpha1]$ kubectl get helmchart
NAME                 CHART     VERSION   SOURCE KIND      SOURCE NAME   READY   STATUS                                         AGE
default-my-podinfo   podinfo   *         HelmRepository   podinfo       True    Pulled 'podinfo' chart with version '6.1.0'.   56s
[v1alpha1]$ kubectl get helmchart
NAME                 CHART     VERSION   SOURCE KIND      SOURCE NAME   READY   STATUS                                                                                   AGE
default-my-podinfo   podinfo   *         HelmRepository   podinfo       False   failed to retrieve source: HelmRepository.source.toolkit.fluxcd.io "podinfo" not found   64s
[v1alpha1]$ kubectl get helmchart
NAME                 CHART     VERSION   SOURCE KIND      SOURCE NAME   READY   STATUS                                                                                   AGE
default-my-podinfo   podinfo   *         HelmRepository   podinfo       False   failed to retrieve source: HelmRepository.source.toolkit.fluxcd.io "podinfo" not found   69s
```
When this happens flux plugin `GetAvailablePackageDetail`() will raise a NotFound error a lot earier than the line you pointed out. It'll be on line 52 (repo, err := s.getRepoInCluster(ctx, repoName)) call. If you think we should change the way it works, let me know

Comment (absoludity) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go:106]: Yeah, fine as is then. Thanks for checking.

**Pull request merged**